### PR TITLE
Incorporate london, arrow glacier, gray glacier EIPs

### PR DESCRIPTION
**Pull request by alexkroeger**
This PR brings master up to date as of the Gray Glacier hard fork, and thus includes 3 upgrades:

- London
- Arrow Glacier
- Gray Glacier

The London hard fork is described [here](https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/london.md) and includes the following EIPs:

- [EIP-1559: Fee market change for ETH 1.0 chain](https://eips.ethereum.org/EIPS/eip-1559) 
- [EIP-3198: BASEFEE opcode](https://eips.ethereum.org/EIPS/eip-3198) 
- [EIP-3529: Reduction in refunds](https://eips.ethereum.org/EIPS/eip-3529) 
- [EIP-3541: Reject new contracts starting with the 0xEF byte](https://eips.ethereum.org/EIPS/eip-3541) 
- [EIP-3554: Difficulty Bomb Delay to December 1st 2021](https://eips.ethereum.org/EIPS/eip-3554) 

The Arrow Glacier and Gray Glacier hard forks were simply delays to the "difficulty bomb." They are simply parameter updates and addressed as well.

TODO:
- [x] repo maintainers create a `berlin` branch that is a copy of master

### CODE REVIEW

File: Paper.tex [hunk 1]
```diff
@@ -904,7 +978,8 @@ \section{Contract Creation}\label{ch:create}\hypertarget{endow}{}
 F \equiv  & \big( \boldsymbol{\sigma}[a] \neq \varnothing \ \wedge\ \big(\boldsymbol{\sigma}[a]_c \neq \texttt{\small KEC}\big(()\big) \vee \boldsymbol{\sigma}[a]_n \neq 0 \big) \big) \quad \vee \\
 \nonumber &(\boldsymbol{\sigma}^{**} = \varnothing \ \wedge\ \mathbf{o} = \varnothing) \quad \vee \\
 \nonumber &g^{**} < c \quad \vee \\
-\nonumber &\lVert \mathbf{o} \rVert > 24576
+\nonumber &\lVert \mathbf{o} \rVert > 24576 \quad \vee \\
+\nonumber &\mathbf{o}[0] = \texttt{0xef}
```
Comment (alexkroeger) [inline Paper.tex:982]: This is where EIP-3541 is incorporated (reject new contract code starting with the 0xef byte).

It is a failure condition for contract creation transactions.

File: Paper.tex [hunk 2]
```diff
@@ -1047,7 +1122,7 @@ \subsection{Execution Environment}\linkdest{exec_env}
 \begin{itemize}
```
Comment (alexkroeger) [inline Paper.tex:1122]: Note that the execution environment needs to contain the base fee, and it does by virtue of containing the block header I__H, so no update is needed to the list to include the base fee.

File: Paper.tex [hunk 3]
```diff
@@ -338,7 +337,20 @@ \subsection{The Transaction} \label{subsec:transaction}
 $T_{\mathrm{w}} = 27 + T_{\mathrm{y}}$ or $T_{\mathrm{w}} = 2\beta + 35 + T_{\mathrm{y}}$ (see EIP-155 by \cite{EIP-155}).
 \end{description}
 
-Additionally, a contract creation transaction (regardless whether legacy or EIP-2930) contains:
+There are differences in how one's acceptable gas price is specified in type 2 transactions versus type 0 and type 1 transactions. Type 2 transactions take advantage of the gas market improvements introduced in EIP-1559 and thus have the following fields related to gas price:
+
+\begin{description}
+  \item[maxFeePerGas]\linkdest{max_fee_per_gas}{} A scalar value equal to the maximum number of Wei to be paid per unit of \textit{gas} for all computation costs incurred as a result of the execution of this transaction; formally $T_{\mathrm{m}}$.
```
Comment (alexkroeger) [inline Paper.tex:343]: I kept the subscripts for `maxFeePerGas` and `maxPriorityFeePerGas` to a single letter, since the convention of the paper looks to be single letters to denote scalar values within tuples.

File: BRANCHES.md [hunk 4]
```diff
@@ -4,8 +4,9 @@ Each protocol version is specified in `Paper.tex` found in a branch of this repo
...
-| master            | [Berlin](https://github.com/ethereum/eth1.0-specs/blob/master/network-upgrades/mainnet-upgrades/berlin.md)        | Since 12,244,000 until 12,964,999 |
+| TBD               | [Paris](https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/gray-glacier.md)    | 15,537,394 until 17,034,869
+| london            | [Gray Glacier](https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/gray-glacier.md) <br> [Arrow Glacier](https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/arrow-glacier.md) <br> [London](https://github.com/ethereum/eth1.0-specs/blob/master/network-upgrades/mainnet-upgrades/london.md)    |   Since 15,050,000 until 15,537,393 <br> Since 13,773,000 until 15,049,999 <br>Since 12,965,000 until 13,772,999 |
```
Comment (alexkroeger) [inline BRANCHES.md:8]: I decided to point the `Arrow Glacier` and `Gray Glacier` to the same branch (london), as this was done for the `Muir Glacier` fork w.r.t `Istanbul`.

These could be separate PRs/branches, however.

File: Paper.tex [hunk 5]
```diff
@@ -171,7 +171,7 @@ \subsection{Which History?}
 The scheme we use in order to generate consensus is a simplified version of the GHOST protocol introduced by \cite{cryptoeprint:2013:881}. This process is described in detail in section \ref{ch:ghost}.
 
 Sometimes, a path follows a new protocol from a particular height (block number).
-This document describes one version of the protocol, namely the \textit{Berlin} version defined by \cite{BeikoBerlin}.
+This document describes the \textit{London}, \textit{Arrow Glacier}, and \textit{Gray Glacier} versions of the protocol.\footnote{Note that the \textit{Arrow Glacier} and \textit{Gray Glacier} forks only affected the \hyperlink{difficulty_bomb}{"difficulty bomb"} described in section \ref{subsec:The_Block}.}
```
Comment (alexkroeger) [inline Paper.tex:174]: I decided to point the Arrow Glacier and Gray Glacier to the same branch (london), as this was done for the Muir Glacier fork w.r.t Istanbul.

These could be separate PRs/branches, however.

File: BRANCHES.md [hunk 6]
```diff
@@ -4,8 +4,9 @@ Each protocol version is specified in `Paper.tex` found in a branch of this repo
...
-| master            | [Berlin](https://github.com/ethereum/eth1.0-specs/blob/master/network-upgrades/mainnet-upgrades/berlin.md)        | Since 12,244,000 until 12,964,999 |
+| TBD               | [Paris](https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/gray-glacier.md)    | 15,537,394 until 17,034,869
+| london            | [Gray Glacier](https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/gray-glacier.md) <br> [Arrow Glacier](https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/arrow-glacier.md) <br> [London](https://github.com/ethereum/eth1.0-specs/blob/master/network-upgrades/mainnet-upgrades/london.md)    |   Since 15,050,000 until 15,537,393 <br> Since 13,773,000 until 15,049,999 <br> Since 12,965,000 until 13,772,999 |
```
Comment (yperbasis) [inline BRANCHES.md:8]: The branch should be `master` instead of `london`

File: Paper.tex (hunk 5 above)
Comment (yperbasis) [reply Paper.tex:174]: Yeah, it's fine to do them in one go with the London upgrade.

File: README.md [hunk 7]
```diff
@@ -10,9 +10,9 @@ It is a free culture work, licensed under Creative Commons Attribution Share-Ali
 
 ## Repository Currently Outdated
 
-The Yellow Paper is out of date. It reflects the Ethereum specification up to the [Berlin](https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/berlin.md) network upgrade, activated on the Ethereum mainnet at block `12_244_000` (April 2021). 
+The Yellow Paper is out of date. It reflects the Ethereum specification up to the [Gray Glacier](https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/gray-glacier.md) network upgrade, activated on the Ethereum mainnet at block `15_050_000` (August 2021).
```
[the remark this answers is not in the document]
Comment (alexkroeger) [reply README.md]: good catch, fixed!

File: BRANCHES.md [hunk 8]
```diff
@@ -4,8 +4,9 @@ Each protocol version is specified in `Paper.tex` found in a branch of this repo
 
 | Branch            | Version                                                           | Applicable Block Numbers        |
 |-------------------|-------------------------------------------------------------------------------------------------------------------|-----------------------------------|
-| TBD               | [Gray Glacier](https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/gray-glacier.md) <br> [Arrow Glacier](https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/arrow-glacier.md) <br> [London](https://github.com/ethereum/eth1.0-specs/blob/master/network-upgrades/mainnet-upgrades/london.md)        | Since 15,050,000 and onwards <br> Since 13,773,000 until 15,049,999 <br> Since 12,965,000 until 13,772,999 |
-| master            | [Berlin](https://github.com/ethereum/eth1.0-specs/blob/master/network-upgrades/mainnet-upgrades/berlin.md)        | Since 12,244,000 until 12,964,999 |
+| TBD               | [Paris](https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/gray-glacier.md)    | 15,537,394 until 17,034,869
```
[the remark this answers is not in the document]
Comment (alexkroeger) [reply BRANCHES.md]: good catch, fixed!

File: BRANCHES.md (hunk 6 above)
Comment (alexkroeger) [reply BRANCHES.md]: updated

File: Paper.tex [hunk 9]
```diff
@@ -338,7 +337,20 @@ \subsection{The Transaction} \label{subsec:transaction}
...
+
+\begin{description}
+  \item[maxFeePerGas]\linkdest{max_fee_per_gas}{} A scalar value equal to the maximum number of Wei to be paid per unit of \textit{gas} for all computation costs incurred as a result of the execution of this transaction; formally $T_{\mathrm{m}}$.
+  \item[maxPriorityFeePerGas]\linkdest{max_priority_fee_per_gas}{} A scalar value equal to the maximum number of Wei to be paid to the block's fee recipient as an incentive to include the transaction; formally $T_{\mathrm{f}}$.
+\end{description}
+
+In contrast, type 0 and type 1 transactions specify gas price as a single value:
+
+\begin{description}
+  \item[gasPrice]\linkdest{tx_gas_price_T__p}{} A scalar value equal to the number of Wei to be paid per unit of \textit{gas} for all computation costs incurred as a result of the execution of this transaction; formally $T_{\mathrm{p}}$.
```
Comment (yperbasis) [inline Paper.tex:350]: It'd be good to note that for type 0 & 1 transactions effectively Tm = Tf = Tp.

File: Paper.tex [hunk 10]
```diff
@@ -610,6 +627,38 @@ \subsubsection{Block Header Validity}
...
+    {P(H)_{\mathrm{H}}}_{\mathrm{f}} + \nu & \text{if} \quad {P(H)_{\mathrm{H}}}_{\mathrm{g}} > \tau \\
+  \end{cases}
+  \end{equation}
+
+where:
+  \begin{align}
+    \tau &\equiv \frac{{P(H)_{\mathrm{H}}}_{\mathrm{l}}}{\rho} \\
+    \rho &\equiv 2 \\
+    \nu &\equiv \begin{cases}
+      \frac{{P(H)_{\mathrm{H}}}_{\mathrm{f}} \times ({\tau - P(H)_{\mathrm{H}}}_{\mathrm{g}})}{\tau \times \xi} &  \text{if} \quad {P(H)_{\mathrm{H}}}_{\mathrm{g}} < \tau \\
```
Comment (yperbasis) [inline Paper.tex:650]: Because of the way how it spec-ed in EIP-1559 with integer division, the more accurate version is
floor(floor(nominator/tau)/xi)

File: Paper.tex [hunk 11]
```diff
@@ -637,19 +686,26 @@ \subsubsection{Block Header Validity}
 \nonumber& & H_{\mathrm{l}} \geqslant 5000  \quad \wedge \\
```
Comment (yperbasis) [inline Paper.tex:686]: It might be relegated to a footnote, but in the check a couple of lines above that the gas limit may change only by less than ±1/1024 the parent gas limit is multiplied by the elasticity multiplier rho when H<sub>i</sub> = F<sub>London</sub>.

**Review by yperbasis — CHANGES_REQUESTED** [review]

File: Paper.tex (hunk 9 above)
Comment (alexkroeger) [reply Paper.tex:350]: That's not precisely accurate to say for Tf I think, since priority fee is derived from gas price and the base fee.

For type 0 & 1 transactions:
`Tf = Tp - Hf`

But you're right to say Tp is analogous to Tm, but it's a guarantee and not a max.

File: Paper.tex (hunk 10 above)
Comment (alexkroeger) [reply Paper.tex:650]: Yea I think you're right. I think using `floor` for division is taken for granted in the paper, but I think you're right on the order of operations here.

Comment (alexkroeger) [reply Paper.tex:650]: I may try to double check this using live values.

File: Paper.tex (hunk 11 above)
Comment (alexkroeger) [reply Paper.tex:686]: I think the elasticity multiplier is irrelevant when validating the gas limit.

From the EIP:
```
		parent_gas_limit = self.parent(block).gas_limit
...
		# check if the block changed the gas limit too much
		assert block.gas_limit < parent_gas_limit + parent_gas_limit // 1024, 'invalid block: gas limit increased too much'
		assert block.gas_limit > parent_gas_limit - parent_gas_limit // 1024, 'invalid block: gas limit decreased too much'

```

File: Paper.tex (hunk 10 above)
Comment (alexkroeger) [reply Paper.tex]: Confirmed the logic using live values (https://dune.com/queries/3232934?category=canonical) and updated this section.

File: Paper.tex (hunk 11 above)
Comment (yperbasis) [reply Paper.tex:690]: I'm talking about this bit:
```
		# on the fork block, don't account for the ELASTICITY_MULTIPLIER to avoid
		# unduly halving the gas target.
		if INITIAL_FORK_BLOCK_NUMBER == block.number:
			parent_gas_target = self.parent(block).gas_limit
			parent_gas_limit = self.parent(block).gas_limit * ELASTICITY_MULTIPLIER 
```

Comment (alexkroeger) [reply Paper.tex:690]: ah, I missed that detail, thanks, I'll incorporate it

Comment (alexkroeger) [reply Paper.tex:700]: I adjusted it like so:
image

As well as the block header validity function:
image

Comment (alexkroeger) [reply Paper.tex:700]: I didn't adjust the definition of the parent block gas target (tau), since this makes no difference to the base fee calculation, since it's a constant at the London fork block.

File: Paper.tex (hunk 9 above)
Comment (alexkroeger) [reply Paper.tex:350]: I was going back through comments, this is correct to say (Tm = Tf = Tp), and that's how the EIP handles type 0 and 1 via `normalize_transaction`. I added to a footnote to this effect.

image

File: Paper.tex [hunk 12]
```diff
@@ -705,9 +776,27 @@ \subsection{Execution}
...
+  f \equiv \begin{cases}
+  T_{\mathrm{p}} - H_{\mathrm{f}} & \text{if} \; T_{\mathrm{x}} = 0 \lor T_{\mathrm{x}} = 1 \\
+  min(T_{\mathrm{f}}, T_{\mathrm{m}} - H_{\mathrm{f}}) & \text{if} \; T_{\mathrm{x}} = 2 \\
+  \end{cases}
+\end{equation}
+
 The up-front cost $v_0$ is calculated as:
 \begin{equation}
-v_0 \equiv \hyperlink{tx_gas_limit_T__g}{T_{\mathrm{g}}} \hyperlink{tx_gas_price_T__p}{T_{\mathrm{p}}} + \hyperlink{tx_value_T__v}{T_{\mathrm{v}}}
+v_0 \equiv \hyperlink{tx_gas_limit_T__g}{T_{\mathrm{g}}} p + \hyperlink{tx_value_T__v}{T_{\mathrm{v}}}
```
Comment (yperbasis) [inline Paper.tex:799]: It should be `Tm` instead of `p` here: see https://github.com/ethereum/EIPs/pull/3594

File: Paper.tex [hunk 13]
```diff
@@ -718,26 +807,27 @@ \subsection{Execution}
 \linkdest{transaction_nonce}{}T_{\mathrm{n}} & = & \boldsymbol{\sigma}[S(T)]_{\mathrm{n}} \quad \wedge \\
 g_0 & \leqslant & T_{\mathrm{g}} \quad \wedge \\
 v_0 & \leqslant & \boldsymbol{\sigma}[S(T)]_{\mathrm{b}} \quad \wedge \\
+p & \geqslant & H_{\mathrm{f}} \quad \wedge \\
```
[the remark this answers is not in the document]
Comment (alexkroeger) [reply Paper.tex:810]: You're right, those are specified in the EIP.

File: Paper.tex (hunk 12 above)
Comment (alexkroeger) [reply Paper.tex]: updated!

File: Paper.tex (hunk 13 above)
[the remark this answers is not in the document]
Comment (alexkroeger) [reply Paper.tex]: updated!

**Review by yperbasis — APPROVED** [review]
